### PR TITLE
set position and orientation using value instead of setValueAtTime

### PR DESCRIPTION
**Pull request by andreasjeppsson**
This fixes an issue in Chrome where the audio in some cases gets stuck jumping from left to right channel.

Examples:
issue: 
http://labs.plan8.se/panner_test/?audioparam=0
fix: 
http://labs.plan8.se/panner_test/?audioparam=1

### CODE REVIEW

File: src/audio/AudioListener.js [hunk 1]
```diff
@@ -103,15 +103,15 @@ AudioListener.prototype = Object.assign( Object.create( Object3D.prototype ), {
...
-				listener.upZ.setValueAtTime( up.z, this.context.currentTime );
+				listener.positionX.value = position.x;
+				listener.positionY.value = position.y;
+				listener.positionZ.value = position.z;
+				listener.forwardX.value = orientation.x;
+				listener.forwardY.value = orientation.y;
+				listener.forwardZ.value = orientation.z;
+				listener.upX.value = up.x;
+				listener.upY.value = up.y;
+				listener.upZ.value = up.z;
```
Comment (hoch) [inline src/audio/AudioListener.js]: I am not really sure what this is trying to fix. What was the issue and how this can be a solution?

Comment (mrdoob) [inline src/audio/AudioListener.js]: /ping @andreasplan8

**Comment by andreasplan8** [comment]
@hoch @mrdoob
The issue is that I get a flutter/tremolo effect where the audio is not panning smoothly when using setValueAtTime() as you can here in the example.

Setting the value directly on the audioParam seems to fix it. Not sure why.

(I'm testing with Chrome 59.0.3071.86)

**Comment by hoch** [comment]
@andreasplan8 

> flutter/tremolo

This means you're changing the parameter value too fast. The parameter setter (=) in Chrome implicitly applies smoothing (leaky integration or dezipper) over the certain time constant, so those modulation artifacts are smoothed out.

Since the implicit smoothing is actually [deprecated from the spec](https://github.com/WebAudio/web-audio-api/pull/393), I recommend to use the automation method. For example,

```js
// To smooth parameter value over the frame-to-frame interval at 60fps.
const tau = 60/1000;
const later = this.context.currentTime + tau;

listener.positionX.linearRampToValueAtTime(position.x, later);
listener.positionY.linearRampToValueAtTime(position.y, later);
listener.positionZ.linearRampToValueAtTime(position.z, later);
```

More typing, but future-proof.

**Comment by mrdoob** [comment]
> ```js
> const tau = 60/1000;
> ```

Hardcoding this is probably not a good idea?

**Comment by hoch** [comment]
Yeah, my code was just an example. If you have a variable that keeps the frame interval, it is probably good enough.

**Comment by mrdoob** [comment]
Déjà vu. https://github.com/mrdoob/three.js/pull/9845#issuecomment-254294017 😁

**Comment by hoch** [comment]
Heh. I also vaguely remembered we talked about this before. :)

**Comment by andreasplan8** [comment]
@mrdoob @hoch Ok, but isn't it better just setting the value instead of scheduling it?

**Comment by hoch** [comment]
@andreasplan8 It depends on the interval of setting values. If the interval is too coarse, then you'll hear glitches. Also I think what we want here is a smoothing transition between coordinates, thus the linear ramping makes sense. Chrome does the internal interpolation when the setter is used,  which is very similar to the code example above in terms of its behavior, but that is a non-compliant to the spec.

**Comment by hoch** [comment]
@mrdoob Someone needs to pick this up? Or should we just close?

**Comment by Mugen87** [comment]
I would like to work on this if there is room for improvement. At the moment, `three.js`  updates the transformation of panner nodes like this:

https://github.com/mrdoob/three.js/blob/1476110de2e1eb55a77f4a6ff9ae443f4117566e/src/audio/PositionalAudio.js#L112-L113

And this is the code for `AudioListener`:

https://github.com/mrdoob/three.js/blob/1476110de2e1eb55a77f4a6ff9ae443f4117566e/src/audio/AudioListener.js#L110-L127

If tested this setup multiple times in different browser with our web audio examples and it seems to work. Is this approach still the recommended way? If not, what would be the ideal solution?

**Comment by hoch** [comment]
In Chrome, using `setPosition()` and using `position{X,Y,X}.setValueAtTime()` are internally identical. So I wouldn't bother to hand-roll the method as shown above. Also I suspect that Chrome is the only browser that implemented AudioParam in PannerNode.

One thing to think about is: Chrome now does not do the internal smoothing between position. If you want to smooth the transition you'll have to use `linearRampToValueAtTime()` or the other automation methods. So if the transition is too fast, the audio might end up with glitches. This is true theoretically, but I have not seen any bug report on it yet.

I think simply using `setPosition` and `setOrientation` might be good enough until all the browsers have the support for AudioParams in PannerNode.

**Comment by hoch** [comment]
Using a constant end time (now + 20ms) for linear ramping is sort of acceptable as well. For the real-time/interactive application, it will like a leaky integrator. It depends on your application.

Using `setTargetAtTime()` is a theoretically correct answer for this use case, but please keep in mind its end time because it does last forever unless the connected source stops.

**Comment by Mugen87** [comment]
Closing in favor of #14393.

**Pull request closed without merging**